### PR TITLE
feat(KSM-388): add custom fields to all 22 resource types

### PR DESCRIPTION
**Pull request by stas-schaller**
## Summary

Implements user-defined custom fields (`custom` block) across all 22 resource types. Custom fields let users attach arbitrary typed data to any Keeper record from Terraform without schema changes — using a single `value` string for simple types and `jsonencode()` for complex structured types.

### New Features

- **Custom fields on all 22 resources** (KSM-388): `login`, `bank_account`, `bank_card`, `birth_certificate`, `contact`, `database_credentials`, `driver_license`, `encrypted_notes`, `file`, `health_insurance`, `membership`, `passport`, `photo`, `server_credentials`, `software_license`, `ssh_keys`, `ssn_card`, `address`, `pam_database`, `pam_directory`, `pam_machine`, `pam_remote_browser`, `pam_user`
- **43+ Keeper field types supported**: simple types (`text`, `secret`, `url`, `email`, `multiline`, `checkbox`, `date`, `birthDate`, `expirationDate`) and structured types (`phone`, `name`, `address`, `paymentCard`, `bankAccount`, `host`, `securityQuestion`, `keyPair`, `script`). Unknown types fall through to plain string storage with no data loss.
- **Multi-value support**: structured types accept `jsonencode([{...},{...}])` for multiple entries in a single field
- **PAM merge logic** (`pam_machine`, `pam_user`): vault-managed "Private Key Passphrase" custom field is preserved across create/update — user-defined fields are merged around it, never overwriting it
- **`required` and `privacy_screen` round-trip**: `Computed: true` on both fields so imported records with these set don't produce perpetual diffs when HCL omits them

### Bug Fixes

- **Date timezone drift**: `date` custom fields now read back in UTC (was using local time, causing one-day drift for users in negative-offset timezones)
- **`birthDate` / `expirationDate`**: these date variants were falling through to plain text; now handled with the same millisecond → YYYY-MM-DD conversion as `date`
- **`checkbox` value type**: was stored as `[]string` (wrong); now correctly uses `core.Checkbox` with `[]bool`, matching all other SDKs
- **Reference type SDK structs**: `addressRef`, `cardRef`, `fileRef`, `oneTimeCode` were falling through to `core.Text`; now use their correct SDK struct types
- **`paymentCard` perpetual diff** (KSM-888, merged via #80): write path read snake_case keys (`card_number`, etc.) from the user's `jsonencode()` map, but SDK struct JSON tags are camelCase. Keys never matched — card always written empty, read back as `{}`, diffed on every plan. Fix: key lookups now use `cardNumber`, `cardExpirationDate`, `cardSecurityCode`.
- **Non-canonical `checkbox` and `date` inputs cause perpetual diffs** (KSM-889, merged via #81): `"yes"`/`"1"` for checkbox were silently coerced to `false`; RFC3339 for date fields was accepted on write but discarded to YYYY-MM-DD on read. Both caused perpetual diffs. Fix: strict validation — only `"true"`/`"false"` for checkbox, only YYYY-MM-DD for date types. Clear error returned for anything else.

### Docs / Examples

- Add `custom` block schema section to all 22 resource docs
- Create `pam_database.md` and `pam_directory.md` (resources existed without docs — pre-existing gap)
- Add `custom` block example to all 22 example `.tf` files

## Reviewer Guide

**XLARGE PR** (3,100+ lines, 83 files) — most of the size is mechanical repetition. Suggested review order:

### Review carefully (core logic — 2 files)

**`secretsmanager/provider.go`**
- `epochMsToDateUTC()` — extracted helper; see unit tests for the timezone edge case it guards
- `parseJSONItems()` — detects JSON array vs single object, enables multi-value fields
- `customFieldsFromSchema()` — full switch over all Keeper field types; each case constructs the correct SDK struct. Key things to review:
  - `checkbox` case: strict `"true"`/`"false"` validation with error (KSM-889)
  - `date`/`birthDate`/`expirationDate` case: YYYY-MM-DD only, no RFC3339 (KSM-889)
  - `paymentCard` case: camelCase key lookups (`cardNumber`, etc.) — KSM-888
- `getFieldItemsData()` — date type check extended to `birthDate`/`expirationDate`; calls `epochMsToDateUTC`

**`secretsmanager/record_fields.go`**
- `schemaCustomField()` — the TypeList schema for `custom`; `required` and `privacy_screen` have `Computed: true`

### Review the pattern, spot-check a few (22 resource files)

Each resource file has the same 4 additions: (1) `"custom": schemaCustomField()` in schema, (2) create block, (3) read via `getFieldItemsData`, (4) update via `customFieldsToDict` + `RawJson` sync.

Different from the pattern — review these individually:
- `resource_pam_machine.go` — merge logic: preserves "Private Key Passphrase" on update, filters it from read
- `resource_pam_user.go` — same merge logic

### Spot-check (tests, docs, examples)

- `secretsmanager/custom_fields_unit_test.go` — unit tests for `epochMsToDateUTC` (UTC timezone regression) and `parseJSONItems` (single object vs array); no vault required
- `secretsmanager/custom_fields_schema_test.go` — verifies `custom` TypeList with all 5 sub-keys is present in all 22 resource schemas; no vault required
- `secretsmanager/resource_login_custom_fields_test.go` — acceptance tests covering all major type categories, import/Computed regression, paymentCard round-trip (KSM-888), and ExpectError regression for checkbox/date non-canonical inputs (KSM-889)
- `resource_pam_machine_test.go`, `resource_pam_user_test.go` — merge logic tests

Estimated review time: 45–60 min (suggest 2 sessions: core logic first, resource files second)

## Breaking Changes

None. Custom fields are additive — existing configs require no changes.

Notes on corrected behavior (not breaking, since these inputs were already producing wrong results):
- `paymentCard`: use camelCase keys in `jsonencode()` — snake_case keys never worked (perpetual diff)
- `checkbox`: use `"true"` or `"false"` — other values now error instead of silently storing `false`
- `date`/`birthDate`/`expirationDate`: use YYYY-MM-DD — RFC3339 now errors instead of silently causing a perpetual diff

## Related Issues

- Closes #16 (custom field support — opened 2023-03-13)
- Jira: KSM-388
- Jira: KSM-888 (paymentCard key fix, merged via #80)
- Jira: KSM-889 (checkbox/date non-canonical input rejection, merged via #81)

### CODE REVIEW

File: examples/resources/folder.tf [hunk 1]
```diff
@@ -20,6 +20,14 @@ provider "secretsmanager" {
 resource "secretsmanager_folder" "my_folder" {
   parent_uid = "<Parent Folder UID>"
   name       = "<Folder Name>"
+
+  # Custom fields — attach arbitrary typed data to the record
```
Comment (idimov-keeper) [inline examples/resources/folder.tf:24]: AFAIK folders don't have fields (only records) - also check folder.md for extra custom section
Because these don't match the schema they may be rejected by Terraform

File: secretsmanager/custom_fields_schema_test.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	resources := map[string]*schema.Resource{
+		"address":              resourceAddress(),
+		"bank_account":         resourceBankAccount(),
+		"bank_card":            resourceBankCard(),
+		"birth_certificate":    resourceBirthCertificate(),
+		"contact":              resourceContact(),
+		"database_credentials": resourceDatabaseCredentials(),
+		"driver_license":       resourceDriverLicense(),
+		"encrypted_notes":      resourceEncryptedNotes(),
+		"health_insurance":     resourceHealthInsurance(),
```
Comment (idimov-keeper) [inline secretsmanager/custom_fields_schema_test.go:28]: resourceFile() / "file" is missing from the map but it implements `schemaCustomField()` in `resource_file.go`

File: secretsmanager/record_fields.go [hunk 3]
```diff
@@ -1513,3 +1522,49 @@ func schemaUrlField() *schema.Schema {
 		},
 	}
 }
+
+// schemaCustomField returns the schema for user-defined custom fields on a record.
+// Each field has a type, label, and value. The value is always a string:
+// - Simple types (text, multiline, secret, url, email): plain string value
+// - Complex types (phone, name, address, paymentCard): jsonencode() of the nested object
+// - Date: RFC3339 string (e.g. "2024-01-15T00:00:00Z")
```
Comment (idimov-keeper) [inline secretsmanager/record_fields.go:1530]: Which one is the right one here....
secretsmanager/record_fields.go — schemaCustomField() comment (lines 1526–1531)
secretsmanager/provider.go — customFieldsFromSchema comment (lines 365–367)
Comments still say dates accept RFC3339 and/or “RFC3339 or YYYY-MM-DD”, but customFieldsFromSchema only accepts YYYY-MM-DD (time.Parse("2006-01-02", …)), matching KSM-889 and the PR notes — not RFC3339.

File: secretsmanager/provider.go [hunk 4]
```diff
@@ -337,6 +346,329 @@ func getFieldItemsData(recordDict map[string]interface{}, section string) []inte
 	return fis
 }
 
+// parseJSONItems accepts either a single JSON object ({"key":"val"}) or a JSON array
+// ([{"key":"val"},{"key":"val2"}]) and returns a slice of raw JSON messages — one per
+// entry. This lets callers write value = jsonencode({...}) for one entry or
+// value = jsonencode([{...},{...}]) for multiple entries without changing the schema.
+func parseJSONItems(value string) ([]json.RawMessage, error) {
+	if strings.TrimSpace(value)[0] == '[' {
```
Comment (idimov-keeper) [inline secretsmanager/provider.go:354]: `strings.TrimSpace(value)[0]` panics if value is empty or all whitespace. You can guard `value != ""`, but whitespace-only value is non-empty and still panics

File: examples/resources/address.tf [hunk 5]
```diff
@@ -34,6 +34,14 @@ resource "secretsmanager_address" "my_address" {
       zip     = "90003-2334"
     }
   }
+
```
Comment (idimov-keeper) [inline examples/resources/address.tf:37]: is this intentional
_There is no examples/resources/pam_remote_browser.tf; the only pam_remote_browser sample under examples/ is examples/ephemeral-resources/pam_remote_browser.tf, which does not demonstrate the managed resource’s custom block._

File: secretsmanager/provider.go [hunk 6]
```diff
@@ -337,6 +346,329 @@ func getFieldItemsData(recordDict map[string]interface{}, section string) []inte
...
+		case "text", "multiline", "secret", "url", "email", "login", "password", "pinCode",
+			"accountNumber", "licenseNumber":
+			f := &core.Text{KeeperRecordField: base, Required: required, PrivacyScreen: privacyScreen}
+			f.Type = fieldType
+			if value != "" {
+				f.Value = []string{value}
+			}
+			fields = append(fields, f)
+
+		case "checkbox":
```
Comment (idimov-keeper) [inline secretsmanager/provider.go:397]: not sure about this specific field type but every other field type here  seems to have a `PrivacyScreen`

**Review by idimov-keeper — CHANGES_REQUESTED** [review]

File: secretsmanager/provider.go (hunk 6 above)
Comment (stas-schaller) [reply secretsmanager/provider.go:397]: Checked against the GO SDK, looks like checkbox doesn't have it: 
```Go
// record_data.go:564 — no PrivacyScreen
type Checkbox struct {
    KeeperRecordField
    Required bool   `json:"required,omitempty"`
    Value    []bool `json:"value,omitempty"`
}
```

**Review by idimov-keeper — APPROVED** [review]

File: secretsmanager/record_fields.go (hunk 3 above)
Comment (stas-schaller) [reply secretsmanager/record_fields.go:1530]: Should be YYYY-MM-DD. I missed these comments and corrected them

File: examples/resources/address.tf (hunk 5 above)
Comment (stas-schaller) [reply examples/resources/address.tf:37]: Added!
